sympy/crypto/crypto.py

```
@@ -2037,3 +2038,200 @@ def dh_shared_key(key, b):
+    Given a message as an integer i, write i in its
+    bit representation b_0,...,b_n. For each k,
+
+     if b_k = 0:
+        let a_k be a random square
+        (quadratic residue) modulo p * q
+        such that jacobi_symbol(a, p * q) = 1
+     if b_k = 1:
+        let a_k be a random non-square
+        (non-quadratic residue) modulo p * q
+        such that jacobi_symbol(a, p * q) = 1
+
+    return [a_1, a_2,...]
+
+    b_k can be recovered by checking whether or not
+    a_k is a residue. And from the b_k's, the message
+    can be reconstructed.
+
+    The idea is that, while jacobi_symbol(a, p * q)
+    can be easily computed (and when it is equal to -1 will
+    tell you that a is not a square mod p * q), quadratic
+    residuosity modulo a composite number is hard to compute
+    without knowing its factorization.
+
+    Moreover, approximately half the numbers coprime to p * q have
+    jacobi_symbol equal to 1. And among those, approximately half
+    are residues and approximately half are not. This maximizes the
+    entropy of the code.
+
+    Parameters
+    ==========
+
+    p, q, a : initialization variables
+
+    Returns
+    =======
+
+    p, q | false (int)
+        if inputs cannot be used to initalize public key
+        then refurn false. Otherwise, return p and q
+
+    """
+    if p == q:
+        return False
+    elif not isprime(p) or not isprime(q):
+        return False
+    elif p == 2 or q == 2:
+        return False
+    return p, q
+
+
+def gm_public_key(p, q, a=None, seed=None):
+    """
+    Compute public keys for p and q.
+    Note that in Goldwasser-Micali Encrpytion,
+    public keys are randomly selected.
+
+    Parameters
+    ==========
+
+    p, q, a : (int) initialization variables
+
+    Returns
+    =======
+
+    a | false (int)
+        if a is not None otherwise some random integer
+        coprime to p and q. returns false
+        if inputs cannot be used to initalize public key
```

It would be better to raise an exception instead of returning False.

---

This wasn't done.